pkg/image/registryclient/client.go

```diff
@@ -223,7 +248,48 @@ func (c *Context) Repository(ctx context.Context, registry *url.URL, repoName st
 	if limiter == nil {
 		limiter = rate.NewLimiter(rate.Limit(5), 5)
 	}
-	return NewLimitedRetryRepository(repo, c.Retries, limiter), nil
+	retryRepo := NewLimitedRetryRepository(repo, c.Retries, limiter)
+	if c.AlternateRepositoryStrategy != nil {
+		retryRepo.alternateStrategy = c.AlternateRepositoryStrategy
+	}
+	return retryRepo, nil
+}
+
+// RepositoryWithAlternateReference returns the next cached repository and ref, or will gather alternate sources then return first repo, ref, or error
+// Once the repository is returned it is removed from the cache.
+func (c *Context) RepositoryWithAlternateReference(ctx context.Context, locator imagereference.DockerImageReference, cachedRepos map[imagereference.DockerImageReference]distribution.Repository) (distribution.Repository, imagereference.DockerImageReference, error) {
+	repo, err := c.repository(ctx, locator.DockerClientDefaults().RegistryURL(), locator.RepositoryName(), true)
```

Comment:
As a side question: are we always consider "insecure = true" for alternate registries?

Reply:
if so, I'll carry the `insecure` variable through from the caller - I'm not sure, I recall a discussion where we decided we can default to true but I might be mis-remembering. 
@smarterclayton can `insecure` be true always here?

Reply:
Image registry only uses "insecure = true" when:

1. ImageStream has an Annotation (`openshift.io/image.InsecureRepository`) set to "true" 
2. The tag being pulled has its `importPolicy.Insecure` set to `true`

As far as I understand this `insecure` argument makes the registry client use the insecure transport and, in some cases, fallback to http instead of https. If it is always true then a dozen tests in image-registry fail (the ones expecting it not to be able to pull from an insecure registry).

Can we have the `insecure` parameter also in `RepositoryWithAlternateReference()`.

Reply:
yes, I will add that, got it

Reply:
added, pushed